app/models/publishable.rb

```diff
@@ -97,12 +101,66 @@ def publish_to_portals(self_url)
 
   def self.included(clazz)
     clazz.class_eval do
-      after_update :auto_publish_to_portal
       has_many :portal_publications, :as => :publishable, :order => :updated_at
+
+      # all changes will be queued for auto publishing
+      after_update :queue_auto_publish_to_portal
+
+      def queue_auto_publish_to_portal
+
+        # set to true when at least one job is added
+        process_queue = false
+
+        urls = self.portal_publications.where(:success => true).pluck(:portal_url).uniq
+        urls.map { |url| Concord::AuthPortal.portal_for_publishing_url(url)}.each do |portal|
+          # no portals are defined in test mode
+          return if portal == nil
+
+          # add the publish job to the table with a unique constraint of portal_url+publishable_id+publishable_type
+          # if there is an existing job for that tuple Rails will throw a StatementInvalid exception which we will ignore
+          # since the existing queued job in the table will take care of this change
+          begin
+            job = PortalPublicationJob.new :portal_url => portal.publishing_url
+            job.publishable = self
+            job.save! # this will throw ActiveRecord::StatementInvalid on duplicate jobs
+            process_queue = true
+          rescue ActiveRecord::StatementInvalid => e
+            raise e unless /Duplicate entry/.match(e.to_s)
+          end
+        end
+
+        # call the delayed job to process the queue if we added at least one job
+        if process_queue
+           # :auto_publish_url is set in application_controller so that the model has access
+          process_auto_publish_to_portal_queue(Thread.current[:auto_publish_url])
+        end
+      end
+
+      def process_auto_publish_to_portal_queue(self_url)
+        json_cache = {}
+
+        PortalPublicationJob.all.each do |job|
+          # immediately remove the job so that new edits can be queued
+          PortalPublicationJob.find(job.id).delete
+
+          # get the json and store it in a local cache in case there are multiple portals
+          cache_key = "#{job.publishable.class.name}:#{job.publishable.id}"
+          json = json_cache[cache_key] || job.publishable.serialize_for_portal(self_url).to_json
+          json_cache[cache_key] = json
+
+          # skip autopublish if the hash hasn't changed since the last publish
+          return if publication_hash == Digest::SHA1.hexdigest(json)
```

Comment:
Where is this publication_hash coming from? It seems it needs to come from the last PortalPublication for this combination of portal and publishable.